Stratis.Bitcoin/BlockPulling/BlockPuller.cs

```diff
@@ -26,14 +64,41 @@ public abstract class BlockPuller : IBlockPuller
         /// <summary>Minimal quality score of a peer node based on the node's past experience with the peer node.</summary>
         public const int MinQualityScore = 1;
 
-        /// <summary>List of relations to peer nodes mapped by block header hashes that the peers are requested to provide.</summary>
-        private readonly ConcurrentDictionary<uint256, BlockPullerBehavior> map;
+        /// <summary>Instance logger.</summary>
+        protected readonly ILogger logger;
+
+        /// <summary>Lock protecting access to <see cref="map"/>, <see cref="pendingInventoryVectors"/>, <see cref="downloadedBlocks"/>, and <see cref="pendingDownloads"/></summary>
+        private readonly object lockObject = new object();
+
+        /// <summary>
+        /// List of relations to peer nodes mapped by block header hashes that the peers are requested to provide.
+        /// </summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<uint256, BlockPullerBehavior> map;
 
         /// <summary>List of block header hashes that the node wants to obtain from its peers.</summary>
-        private readonly ConcurrentBag<uint256> pendingInventoryVectors;
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Queue<uint256> pendingInventoryVectors;
 
         /// <summary>List of unprocessed downloaded blocks mapped by their header hashes.</summary>
-        protected readonly ConcurrentDictionary<uint256, DownloadedBlock> DownloadedBlocks;
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<uint256, DownloadedBlock> downloadedBlocks;
+
+        /// <summary>Number of items in <see cref="downloadedBlocks"/>. This is for statistical purposes only.</summary>
+        public int DownloadedBlocksCount
+        {
+            get
+            {
+                lock (this.lockObject)
+                {
+                    return this.downloadedBlocks.Count;
+                }
+            }
+        }
+
+        /// <summary>Hash set holding set of block header hashes that are being downloaded, mapped by connected peer's behavior.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<BlockPullerBehavior, HashSet<uint256>> pendingDownloads = new Dictionary<BlockPullerBehavior, HashSet<uint256>>();
```

Comment:
depends on what name we end up for the "map"
if it is "assignedBlockTasks" as i suggest, it would be "peersTasks" here
but if we go with yours "blocksToPeers" then yes, it would be "peersToBlocks" here, i just don't know how i feel about that "X-To-Y" naming for dictionaries